### DISCUSSION TITLE
Issue with Dependency Injection with Shell

### DISCUSSION THREAD

**Discussion by matt-goldman** [dotnet/maui#5263] [Q&A]
I've encountered an issue which I think may be a common scenario, so hoping someone has some advice for me.

I'm building a demo app using Shell, and my Shell has a MenuItem for logging out. I think this is a fairly common thing people may want in an app. To facilitate this, I am injecting a ViewModel into my Shell, and a service handling authentication into my ViewModel.

Because of this, I need to inject my Shell into the constructor of my App class (rather than doing `MainPage = new AppShell();`). But here's the problem - in my Shell's XAML, I want to use colours defined in the resource dictionary of my App. **But now I have a chicken and egg problem**, because I'm trying to consume them in something that's getting injected into the constructor of the class where they are declared. So the app fails to start and throws an error when trying to construct the Shell (that the resources aren't defined - which makes sense as they are not defined yet).

I can think of a few hacky ways to get around this. But what do people think would be a clean approach?

**Comment by jfversluis** [comment]
@PureWeen you have any thoughts on this? Is this something we're aware of?

**Comment by matt-goldman** [comment]
This is my workaround for now:

```csharp
public App(ShellViewModel shellViewModel)
{
    InitializeComponent();

    MainPage = new AppShell(shellViewModel);
}
```

The other approaches I thought of involved using Activator or resolving from the service collection within the App class. This approach felt the least 'wrong' to me.

**Comment by matt-goldman** [comment]
Hey @jfversluis and @PureWeen any thoughts on this? I'm encountering this in a real-world scenario now. Should I convert this to an issue?

**Reply by jfversluis** [reply]
If you think this is something that is currently not possible and/or should be improved probably open an issue! Make sure there isn't one already :)

**Reply by matt-goldman** [reply]
There is this issue: https://github.com/dotnet/maui/issues/3878 This I think is the same problem. The suggestion there is to use `DynamicResource` instead of `StaticResource`; I'll test that and confirm if that works in my scenario.

If it does, and that's the recommended approach (i.e. using the app's dictionaries when injecting Shell or pages into the `App` class), would be cool if that were documented somewhere (or maybe as a tip in a video? ;) )

**Reply by matt-goldman** [reply]
Yep confirmed. Got a repo here demonstrating it: https://github.com/matt-goldman/ResourceDictionaryWithDI.git

In `MainPage.xaml`, the `BackgroundColor` property of a `Button` is set using `StaticResource` calling a value defined in `Colors.xaml`. This works, but when I constructor inject that page into `App.xaml.cs`, it throws the expected error. Changing it to `DynamicResource` gives the expected result (no error, and it loads the color from the dictionary once the dictionary is merged).
